pkg/cloud/gcp/actuators/machine/reconciler.go

```diff
@@ -492,20 +492,26 @@ func (r *Reconciler) exists() (bool, error) {
 			// this error type bubbles back up to the machine-controller to allow
 			// us to delete machines that were never properly created due to
 			// invalid zone.
-			return false, machinecontroller.InvalidMachineConfiguration(fmt.Sprintf("%s: Machine does not exist", r.machine.Name))
+
+			// If the machine has a node assigned, we know that it actually exists,
+			// but it also has an invalid configuration.
+			if r.machine.Status.NodeRef != nil {
```

Comment:
NodeRef isn't good enough, NodeRef only comes once the node has joined the cluster. If you use providerID instead, this will be set as soon as the VM is created on the cloud provider, which is earlier than the NodeRef.

There is a period between the VM being created and the providerID being set that the current code would cause an issue